// simple-uniswap-sdk/dist/factories/pair/uniswap-pair.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniswapPair = void 0;
const ethers_1 = require("ethers");
const error_codes_1 = require("../../common/errors/error-codes");
const uniswap_error_1 = require("../../common/errors/uniswap-error");
const is_address_1 = require("../../common/utils/is-address");
const ethers_provider_1 = require("../../ethers-provider");
const tokens_factory_1 = require("../token/tokens.factory");
const uniswap_pair_settings_1 = require("./models/uniswap-pair-settings");
const uniswap_pair_factory_1 = require("./uniswap-pair.factory");
class UniswapPair {
    constructor(_uniswapPairContext) {
        this._uniswapPairContext = _uniswapPairContext;
        if (!this._uniswapPairContext.fromTokenContractAddress) {
            throw new uniswap_error_1.UniswapError('Must have a `fromTokenContractAddress` on the context', error_codes_1.ErrorCodes.fromTokenContractAddressRequired);
        }
        if (!is_address_1.isAddress(this._uniswapPairContext.fromTokenContractAddress)) {
            throw new uniswap_error_1.UniswapError('`fromTokenContractAddress` is not a valid contract address', error_codes_1.ErrorCodes.fromTokenContractAddressNotValid);
        }
        this._uniswapPairContext.fromTokenContractAddress = ethers_1.ethers.utils.getAddress(this._uniswapPairContext.fromTokenContractAddress);
        if (!this._uniswapPairContext.toTokenContractAddress) {
            throw new uniswap_error_1.UniswapError('Must have a `toTokenContractAddress` on the context', error_codes_1.ErrorCodes.toTokenContractAddressRequired);
        }
        if (!is_address_1.isAddress(this._uniswapPairContext.toTokenContractAddress)) {
            throw new uniswap_error_1.UniswapError('`toTokenContractAddress` is not a valid contract address', error_codes_1.ErrorCodes.toTokenContractAddressNotValid);
        }
        this._uniswapPairContext.toTokenContractAddress = ethers_1.ethers.utils.getAddress(this._uniswapPairContext.toTokenContractAddress);
        if (!this._uniswapPairContext.ethereumAddress) {
            throw new uniswap_error_1.UniswapError('Must have a `ethereumAddress` on the context', error_codes_1.ErrorCodes.ethereumAddressRequired);
        }
        if (!is_address_1.isAddress(this._uniswapPairContext.ethereumAddress)) {
            throw new uniswap_error_1.UniswapError('`ethereumAddress` is not a valid address', error_codes_1.ErrorCodes.ethereumAddressNotValid);
        }
        this._uniswapPairContext.ethereumAddress = ethers_1.ethers.utils.getAddress(this._uniswapPairContext.ethereumAddress);
        const chainId = this._uniswapPairContext
            .chainId;
        const providerUrl = (this._uniswapPairContext).providerUrl;
        if (providerUrl && chainId) {
            this._ethersProvider = new ethers_provider_1.EthersProvider(chainId, providerUrl);
            return;
        }
        if (chainId) {
            this._ethersProvider = new ethers_provider_1.EthersProvider(chainId);
            return;
        }
        throw new uniswap_error_1.UniswapError('You must have a chainId on the context.', error_codes_1.ErrorCodes.youMustSupplyAChainId);
    }
    /**
     * Create factory to be able to call methods on the 2 tokens
     */
    createFactory() {
        return __awaiter(this, void 0, void 0, function* () {
            const tokensFactory = new tokens_factory_1.TokensFactory(this._ethersProvider);
            const tokens = yield tokensFactory.getTokens([
                this._uniswapPairContext.fromTokenContractAddress,
                this._uniswapPairContext.toTokenContractAddress,
            ]);
            const uniswapFactoryContext = {
                fromToken: tokens.find((t) => t.contractAddress ===
                    this._uniswapPairContext.fromTokenContractAddress),
                toToken: tokens.find((t) => t.contractAddress === this._uniswapPairContext.toTokenContractAddress),
                ethereumAddress: this._uniswapPairContext.ethereumAddress,
                settings: this._uniswapPairContext.settings || new uniswap_pair_settings_1.UniswapPairSettings(),
                ethersProvider: this._ethersProvider,
            };
            return new uniswap_pair_factory_1.UniswapPairFactory(uniswapFactoryContext);
        });
    }
}
exports.UniswapPair = UniswapPair;
