// simple-uniswap-sdk/dist/__TEST-SCRIPT__/playground.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const chain_id_1 = require("../enums/chain-id");
const uniswap_version_1 = require("../enums/uniswap-version");
const uniswap_pair_settings_1 = require("../factories/pair/models/uniswap-pair-settings");
const uniswap_pair_1 = require("../factories/pair/uniswap-pair");
// WBTC - 0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599
// FUN - 0x419D0d8BdD9aF5e606Ae2232ed285Aff190E711b
// REP - 0x1985365e9f78359a9B6AD760e32412f4a445E862
// WETH - 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2
// UNI - 0x1f9840a85d5af5bf1d1762f925bdaddc4201f984
// AAVE - 0x7fc66500c84a76ad7e9c93437bfc5ac33e2ddae9
const routeTest = () => __awaiter(void 0, void 0, void 0, function* () {
    console.log(new Date().getTime());
    const fromTokenContractAddress = '0x7fc66500c84a76ad7e9c93437bfc5ac33e2ddae9'; //'0xEf0e839Cf88E47be676E72D5a9cB6CED99FaD1CF';
    const toTokenContractAddress = '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2'; // 0x1985365e9f78359a9B6AD760e32412f4a445E862
    const ethereumAddress = '0xB1E6079212888f0bE0cf55874B2EB9d7a5e02cD9';
    const uniswapPair = new uniswap_pair_1.UniswapPair({
        fromTokenContractAddress,
        toTokenContractAddress,
        ethereumAddress,
        chainId: chain_id_1.ChainId.MAINNET,
        settings: new uniswap_pair_settings_1.UniswapPairSettings({
            // if not supplied it use `0.005` which is 0.5%;
            // all figures
            slippage: 0.005,
            // if not supplied it will use 20 a deadline minutes
            deadlineMinutes: 20,
            disableMultihops: true,
            uniswapVersions: [uniswap_version_1.UniswapVersion.v3],
        }),
    });
    const uniswapPairFactory = yield uniswapPair.createFactory();
    // try {
    const trade = yield uniswapPairFactory.trade('10');
    console.log(trade);
    // console.log(
    //   trade.allTriedRoutesQuotes.filter(
    //     (c) => c.uniswapVersion === UniswapVersion.v3
    //   )
    // );
    console.log(new Date().getTime());
    // } catch (error) {
    //   console.log(error.message);
    // }
    process.stdin.resume();
    // console.log(JSON.stringify(trade));
    // const data = await uniswapPairFactory.generateApproveMaxAllowanceData();
    // console.log(data);
    // const toToken = uniswapPairFactory.toToken;
    // console.log(toToken);
    // const fromToken = uniswapPairFactory.fromToken;
    // console.log(fromToken);
    // const tokenContractAddress = '0x419D0d8BdD9aF5e606Ae2232ed285Aff190E711b';
    // const tokenFactoryPublic = new TokenFactoryPublic(
    //   fromTokenContractAddress,
    //   ChainId.MAINNET
    // );
    // console.log(
    //   await tokenFactoryPublic.getAllowanceAndBalanceOf(ethereumAddress)
    // );
    // // the contract address for which you are allowing to move tokens on your behalf
    // const spender = '0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D';
    // // the amount you wish to allow them to move, this example just uses the max
    // // hex. If not each time they do a operation which needs to move tokens then
    // // it will cost them 2 transactions, 1 to approve the allowance then 1 to actually
    // // do the contract call to move the tokens.
    // const value =
    //   '0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff';
    // const data = tokenFactoryPublic.generateApproveAllowanceData(spender, value);
    // console.log(data);
});
routeTest();
