// simple-uniswap-sdk/dist/factories/uniswap-factory/v2/uniswap-contract.factory.v2.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniswapContractFactoryV2 = void 0;
const uniswap_contract_context_v2_1 = require("../../../uniswap-contract-context/uniswap-contract-context-v2");
class UniswapContractFactoryV2 {
    constructor(_ethersProvider) {
        this._ethersProvider = _ethersProvider;
        this._uniswapFactoryContract = this._ethersProvider.getContract(JSON.stringify(uniswap_contract_context_v2_1.UniswapContractContextV2.factoryAbi), uniswap_contract_context_v2_1.UniswapContractContextV2.factoryAddress);
    }
    allPairs(parameter0) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this._uniswapFactoryContract.allPairs(parameter0);
        });
    }
    allPairsLength() {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this._uniswapFactoryContract.allPairsLength()).toHexString();
        });
    }
    createPair(tokenA, tokenB) {
        return this._uniswapFactoryContract.interface.encodeFunctionData('createPair', [tokenA, tokenB]);
    }
    getPair(token0, token1) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this._uniswapFactoryContract.getPair(token0, token1);
        });
    }
    feeTo() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this._uniswapFactoryContract.feeTo();
        });
    }
    feeToSetter() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this._uniswapFactoryContract.feeToSetter();
        });
    }
    setFeeTo(_feeTo) {
        return __awaiter(this, void 0, void 0, function* () {
            return this._uniswapFactoryContract.interface.encodeFunctionData('setFeeTo', [_feeTo]);
        });
    }
    setFeeToSetter(_feeToSetter) {
        return __awaiter(this, void 0, void 0, function* () {
            return this._uniswapFactoryContract.interface.encodeFunctionData('setFeeToSetter', [_feeToSetter]);
        });
    }
}
exports.UniswapContractFactoryV2 = UniswapContractFactoryV2;
